samza-core/src/main/java/org/apache/samza/control/EndOfStreamManager.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.control;
+
+import com.google.common.collect.Multimap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.apache.samza.message.EndOfStreamMessage;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.StreamMetadataCache;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.task.MessageCollector;
+import org.apache.samza.task.TaskCoordinator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class handles the end-of-stream control message. It aggregates the end-of-stream state for each input ssps of
+ * a task, and propagate the eos messages to downstream intermediate streams if needed.
+ */
+public class EndOfStreamManager {
+  private static final Logger log = LoggerFactory.getLogger(EndOfStreamManager.class);
+
+  private final String taskName;
+  private final MessageCollector collector;
+  // end-of-stream state per ssp
+  private final Map<SystemStreamPartition, EndOfStreamState> eosStates;
+  private final StreamMetadataCache metadataCache;
+  // mapping from input stream to its downstream tasks
+  private final Multimap<SystemStream, String> inputToTasks;
+
+  // topology information. Set during init()
+  private IOGraph ioGraph;
+  // mapping from output stream to its upstream task count
+  private Map<SystemStream, Integer> upstreamTaskCounts;
+
+  public EndOfStreamManager(String taskName,
+      Multimap<SystemStream, String> inputToTasks,
+      Set<SystemStreamPartition> ssps,
+      StreamMetadataCache metadataCache,
+      MessageCollector collector) {
+    this.taskName = taskName;
+    this.inputToTasks = inputToTasks;
+    this.metadataCache = metadataCache;
+    this.collector = collector;
+    Map<SystemStreamPartition, EndOfStreamState> states = new HashMap<>();
+    ssps.forEach(ssp -> {
+        states.put(ssp, new EndOfStreamState());
+      });
+    this.eosStates = Collections.unmodifiableMap(states);
+  }
+
+  public void init(IOGraph ioGraph) {
+    this.ioGraph = ioGraph;
+    this.upstreamTaskCounts = ControlMessageUtils.calculateUpstreamTaskCounts(inputToTasks, ioGraph);
+  }
+
+  public void update(IncomingMessageEnvelope envelope, TaskCoordinator coordinator) {
+    EndOfStreamState state = eosStates.get(envelope.getSystemStreamPartition());
+    EndOfStreamMessage message = (EndOfStreamMessage) envelope.getMessage();
+    state.update(message.getTaskName(), message.getTaskCount());
```

Comment:
Yes, it's done before the task can kick off any threads/async process. So it's already single threaded.

I created https://issues.apache.org/jira/browse/SAMZA-1352 for your second suggestion. I will submit another pr for that.